src/test/java/oripa/domain/cptool/LineAdderTest.java

```diff
@@ -34,40 +37,131 @@
+		adder = new LineAdder();
+	}
+
+	@Test
+	void should_split_existing_lines_if_new_line_crosses_them() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(DIAGONAL_LINE));
+		var line = new OriLine(200, 0, -200, 0, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(4, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(-200, -200, 0, 0, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 200, 200, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(200, 0, 0, 0, OriLine.Type.VALLEY)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, -200, 0, OriLine.Type.VALLEY)));
+	}
+
+	@Test
+	void should_remove_duplicated_segments_if_two_lines_of_same_type_are_overlapping() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(new OriLine(0, 0, 200, 0, OriLine.Type.VALLEY)));
+		var line = new OriLine(0, 0, 100, 0, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(2, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 100, 0, OriLine.Type.VALLEY)));
+		assertTrue(creasePattern.contains(new OriLine(100, 0, 200, 0, OriLine.Type.VALLEY)));
+	}
+
+	@Test
+	void should_split_existing_lines_if_new_line_touches_them() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(DIAGONAL_LINE));
+		var line = new OriLine(200, 0, 0, 0, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(3, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(-200, -200, 0, 0, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 200, 200, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(200, 0, 0, 0, OriLine.Type.VALLEY)));
+	}
+
+	@Test
+	void should_replace_overlapping_parts() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(new OriLine(-200, -200, 0, 0, OriLine.Type.MOUNTAIN)));
+		var line = new OriLine(-100, -100, 100, 100, OriLine.Type.VALLEY);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(3, creasePattern.size());
+		assertTrue(creasePattern.contains(new OriLine(-200, -200, -100, -100, OriLine.Type.MOUNTAIN)));
+		assertTrue(creasePattern.contains(new OriLine(-100, -100, 0, 0, OriLine.Type.VALLEY)));
+		assertTrue(creasePattern.contains(new OriLine(0, 0, 100, 100, OriLine.Type.VALLEY)));
+	}
+
+	@Test
+	void should_remove_multiple_fully_overlapping_line_with_splits() {
+		// Given
+		var creasePattern = new ArrayList<>(List.of(
+				new OriLine[] { new OriLine(-100, 0, -50, 0, OriLine.Type.VALLEY),
+						new OriLine(0, 0, 50, 0, OriLine.Type.VALLEY),
+						new OriLine(100, 0, 150, 0, OriLine.Type.VALLEY) }));
+
+		var line = new OriLine(-200, 0, 200, 0, OriLine.Type.MOUNTAIN);
+
+		// When
+		adder.addLine(line, creasePattern);
+
+		// Then
+		assertEquals(7, creasePattern.size());
```

Comment:
According to the result of test, this line fails and `LineAdder::addLine` is not correct.

Reply:
Yes, that corresponds to in program testing I have done, not sure if this relates to not having an EPS for `OriLine.equals` since lines get removed from a list utilizes that.

Reply:
lifesbest23@66effa7321b07a0ada28ce818f833ec80453bb3a should resolve this.